src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs

```diff
@@ -2306,6 +2311,620 @@ internal unsafe void CM(B b)
+    static void M1()
+    {
+        delegate*<string, string, void> ptr = &M;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (8,48): error CS0121: The call is ambiguous between the following methods or properties: 'C.M(string, object)' and 'C.M(object, string)'
+                //         delegate*<string, string, void> ptr = &M;
+                Diagnostic(ErrorCode.ERR_AmbigCall, "M").WithArguments("C.M(string, object)", "C.M(object, string)").WithLocation(8, 48)
+            );
+        }
+
+        [Fact]
+        public void AddressOf_AsLvalue()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    static void M() {}
+    static void M1()
+    {
+        delegate*<void> ptr = &M;
+        &M = ptr;
+        M2(&M);
+        ref delegate*<void> ptr2 = ref &M;
+    }
+    static void M2(ref delegate*<void> ptr) {}
+}");
+
+            comp.VerifyDiagnostics(
+                // (8,9): error CS0131: The left-hand side of an assignment must be a variable, property or indexer
+                //         &M = ptr;
+                Diagnostic(ErrorCode.ERR_AssgLvalueExpected, "&M").WithLocation(8, 9),
+                // (9,12): error CS1503: Argument 1: cannot convert from '&method group' to 'ref delegate*<void>'
+                //         M2(&M);
+                Diagnostic(ErrorCode.ERR_BadArgType, "&M").WithArguments("1", "&method group", "ref delegate*<void>").WithLocation(9, 12),
+                // (10,40): error CS1510: A ref or out value must be an assignable variable
+                //         ref delegate*<void> ptr2 = ref &M;
+                Diagnostic(ErrorCode.ERR_RefLvalueExpected, "&M").WithLocation(10, 40)
+            );
+        }
+
+        [Fact]
+        public void AddressOf_MethodParameter()
+        {
+            var verifier = CompileAndVerifyFunctionPointers(@"
+using System;
+unsafe class C
+{
+    static void M(string s) => Console.Write(s);
+    static void Caller(delegate*<string, void> ptr) => ptr(""1"");
+    static void Main()
+    {
+        Caller(&M);
+    }
+}", expectedOutput: "1");
+
+            verifier.VerifyIL("C.Main()", expectedIL: @"
+{
+  // Code size       12 (0xc)
+  .maxstack  1
+  IL_0000:  ldftn      ""void C.M(string)""
+  IL_0006:  call       ""void C.Caller(delegate*<string,void>)""
+  IL_000b:  ret
+}
+");
+        }
+
+        [Fact]
+        public void AddressOf_CannotAssignToVoidStar()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    static void M()
+    {
+        void* ptr = &M;
+    }
+}");
+
+            comp.VerifyDiagnostics(
+                // (6,21): error CS0428: Cannot convert method group 'M' to non-delegate type 'void*'. Did you intend to invoke the method?
+                //         void* ptr = &M;
+                Diagnostic(ErrorCode.ERR_MethGrpToNonDel, "&M").WithArguments("M", "void*").WithLocation(6, 21)
+            );
+        }
```

Comment:
> } [](start = 7, length = 2)

Please add tests for
- generic methods including type inference scenarios
- extension methods